test/integration/transaction.test.js

```diff
@@ -414,73 +414,105 @@ if (current.dialect.supports.transactions) {
-                  } catch (e) {
-                    console.log(e.message);
-                  }
-
-                  await Task.create({ id: 2 }, { transaction });
+                    }
+                  );
                 } catch (e) {
                   console.log(e.message);
                 }
 
-                throw new Error('Rollback!');
-              })
-              .catch(() => {});
+                await Task.create({ id: 2 }, { transaction });
+              } catch (e) {
+                console.log(e.message);
+              }
 
-          await this.sequelize.sync({ force: true });
-          await Task.create({ id: 0 });
-          await Task.create({ id: 1 });
+              throw new Error('Rollback!');
+            })
+            .catch(() => {});
+        };
 
-          await Promise.all([update(1, 0), update(0, 1)]);
+        it('should treat deadlocked transaction as rollback', async function () {
+          const Task = await getAndInitializeTaskModel(this.sequelize);
+
+          // This gets called twice simultaneously, and we expect at least one of the calls to encounter a
+          // deadlock (which effectively rolls back the active transaction).
+          // We only expect createTask() to insert rows if a transaction is active.  If deadlocks are handled
+          // properly, it should only execute a query if we're actually inside a real transaction.  If it does
+          // execute a query, we expect the newly-created rows to be destroyed when we forcibly rollback by
+          // throwing an error.
+          // tl;dr; This test is designed to ensure that this function never inserts and commits a new row.
+          await Promise.all([update(this.sequelize, Task, 1, 0), update(this.sequelize, Task, 0, 1)]);
 
           const count = await Task.count();
           // If we were actually inside a transaction when we called `Task.create({ id: 2 })`, no new rows should be added.
           expect(count).to.equal(2, 'transactions were fully rolled-back, and no new rows were added');
         });
+
+        it('should release the connection for a deadlocked transaction', async function () {
+          const Task = await getAndInitializeTaskModel(this.sequelize);
+
+          // 1 of 2 queries should deadlock and be rolled back by InnoDB
+          this.sinon.spy(this.sequelize.connectionManager, 'releaseConnection');
+          await Promise.all([update(this.sequelize, Task, 1, 0), update(this.sequelize, Task, 0, 1)]);
+
+          // Verify that both of the connections were released
+          expect(this.sequelize.connectionManager.releaseConnection.callCount).to.equal(2);
+
+          // Verify that a follow-up READ_COMMITTED works as expected.
+          // For unknown reasons, we need to explicitly rollback on MariaDB,
+          // even though the transaction should've automatically been rolled
+          // back.
+          // Otherwise, this READ_COMMITTED doesn't work as expected.
```

Comment:
I think these comments belong on the code itself (not the test), do you agree?

Reply:
I agree - I added a longer comment to the code itself to reflect this!

```ts
    } catch (err) {
      // MariaDB automatically rolls-back transactions in the event of a
      // deadlock.
      //
      // Even though we shouldn't need to do this, we initiate a manual
      // rollback. Without the rollback, the connection ends up in a limbo
      // state where future transactions may not behave properly.
      //
      // For example (in our tests), a follow-up READ_COMMITTED transaction
      // doesn't work as expected unless we explicitly rollback the
      // transaction: it would fail to read a value inserted outside of that
      // transaction.
      if (options.transaction && err.errno === 1213) {
        try {
          await options.transaction.rollback();
        } catch (err) {
          // Ignore errors - since MariaDB automatically rolled back, we're
          // not that worried about this redundant rollback failing.
        }
        options.transaction.finished = 'rollback';
      }
```